### PR TITLE
Entity match charges with children if possible

### LINKED ISSUE #1012: charge matching
Because we are ordering db_charges with more information first, we ensure that charges with less information (fewer children) never take the spot of a better match (matches children as well as own info).
```
order db_charges: charges with more children are processed first
for db_charge in db_charges:
  potential_matches_without_comparing_children = get them()
  potential_matches_with_comparing_children = get them()
  if potential_matches_without_comparing_children > 1:
    if potential_matches_with_comparing_children:
       match db_charge to first potential_matches_with_comparing_children
    else:
      match db_charge to first potential_matches_without_comparing_children
```
  cc @terinpw

### PR DESCRIPTION
**Pull request by terinpw**
Fixes #1012 

When possible, we should attempt to match charges based on charge fields and sentence/bond fields. Only if we find no matches while comparing sentence/bond fields do we attempt to match based on the charge fields alone. 

Co Author - jamwalla

### CODE REVIEW

File: recidiviz/persistence/entity_matching.py [hunk 1]
```diff
@@ -287,6 +274,10 @@ def _drop_hold(hold: entities.Hold):
         hold.status = HoldStatus.INFERRED_DROPPED
 
 
+def _charge_relationship_count(charge: entities.Charge):
```
Comment (jessex) [inline recidiviz/persistence/entity_matching.py]: Can you add a small comment explaining exactly what's being counted here? I think it's non-obvious.

File: recidiviz/persistence/entity_matching.py [hunk 2]
```diff
@@ -296,15 +287,24 @@ def match_charges(
     |db_booking|, the primary key is updated on the ingested charge. All
     db charges that are not matched to an ingested charge are marked dropped and
     added to the |ingested_booking|.
-    Args:
-        ingested_booking: (entities.Booking)
-        db_booking: (entities.Booking)
+
+    Note: When matching charges, we attempt to match considering children (
+    bonds and sentences) as well as charge information. If we cannot match while
```
Comment (jessex) [inline recidiviz/persistence/entity_matching.py]: Can we add a bit more detail to this? For example, what happens if two charges look the same but have different bonds or sentences? What happens if they look the same, but the db charge has bonds or sentences and the ingested charge has none, or vice versa? I think we should very much err on the side of over-documenting entity matching, given how critical it is and how complex it is.

File: recidiviz/persistence/entity_matching.py [hunk 3]
```diff
@@ -296,15 +287,24 @@ def match_charges(
...
+    comparing children, we match on just charge information.
     """
     dropped_charges = []
-    for db_charge in db_booking.charges:
-        ingested_charge = _get_next_available_match(db_charge,
-                                                    ingested_booking.charges,
-                                                    utils.is_charge_match)
+    sorted_db_charges = sorted(
```
Comment (zdg2102) [inline recidiviz/persistence/entity_matching.py]: Can you add a comment explaining the impact of sorting here?

File: recidiviz/tests/persistence/entity_matching_test.py [hunk 4]
```diff
@@ -411,6 +411,93 @@ def test_matchCharges(self):
...
+        ingested_charge = attr.evolve(db_charge, charge_id=None, bond=None)
+
+        expected_matched_charge = attr.evolve(
+            ingested_charge, charge_id=db_charge.charge_id)
+
+        db_booking = entities.Booking.new_with_defaults(charges=[db_charge])
+        ingested_booking = entities.Booking.new_with_defaults(
+            charges=[ingested_charge])
+
+        # import ipdb; ipdb.set_trace()
```
Comment (zdg2102) [inline recidiviz/tests/persistence/entity_matching_test.py]: Is this line supposed to be here?

File: recidiviz/tests/persistence/entity_matching_test.py [hunk 5]
```diff
@@ -411,6 +411,93 @@ def test_matchCharges(self):
...
+            db_booking=db_booking, ingested_booking=ingested_booking)
+        self.assertCountEqual(ingested_booking.charges,
+                              [expected_matched_charge])
+
+    def test_matchCharges_disambiguateByChildren(self):
+        db_bond = entities.Bond.new_with_defaults(
+            bond_id=_ID, external_id=_EXTERNAL_ID_ANOTHER,
+            status=BondStatus.PENDING)
+        db_bond_another = entities.Bond.new_with_defaults(
+            bond_id=_ID_ANOTHER, external_id=_EXTERNAL_ID,
```
Comment (zdg2102) [inline recidiviz/tests/persistence/entity_matching_test.py]: I might not be reading this right, but wouldn't these disambiguate by having different external IDs without even needing to look at their children?

Comment (terinpw) [reply recidiviz/tests/persistence/entity_matching_test.py]: This is in the bond though, which is the child. So the charges are the same, so only when comparing the bonds do we know how to match.

File: recidiviz/tests/persistence/entity_matching_test.py (hunk 4 above)
Comment (terinpw) [reply recidiviz/tests/persistence/entity_matching_test.py]: Nah, removed it.

File: recidiviz/tests/persistence/entity_matching_test.py (hunk 5 above)
Comment (zdg2102) [reply recidiviz/tests/persistence/entity_matching_test.py]: Oh whoops... my bad

File: recidiviz/persistence/entity_matching.py (hunk 1 above)
Comment (terinpw) [reply recidiviz/persistence/entity_matching.py]: Good idea, done

File: recidiviz/persistence/entity_matching.py (hunk 2 above)
Comment (terinpw) [reply recidiviz/persistence/entity_matching.py]: Yeah great point. Can you take a look and let us know if you think the doc should be updated?

File: recidiviz/persistence/entity_matching.py (hunk 3 above)
Comment (terinpw) [reply recidiviz/persistence/entity_matching.py]: Addressed this in the docstring above - I think that should suffice, but let us know if you think it's still not clear. 

Also updated the variable name to be more descriptive.

Comment (zdg2102) [reply recidiviz/persistence/entity_matching.py]: Makes sense to me

File: recidiviz/persistence/entity_matching.py [hunk 6]
```diff
@@ -296,15 +288,55 @@ def match_charges(
...
+    be marked dropped.
+
+    Because we can match two charges whose children are not equal, we sort our
+    db_charges, attempting to match those with more children before those with
+    fewer or none.
+
+    Examples:
+        1. Two identical charges are in our DB, one with a bond (A) and one
+        without (B). The same data is scraped the next day so that we have
+        corresponding ingested bonds C (with a bond) and D (without a bond).
```
Comment (jessex) [inline recidiviz/persistence/entity_matching.py]: "bonds C" should be "charges C", no?

Comment (terinpw) [reply recidiviz/persistence/entity_matching.py]: Good catch!

**Pull request merged**